sdk/python/kfp/dsl/pipeline_channel.py

```diff
@@ -265,6 +284,234 @@ def __init__(
+            f'{self.condition_branches_group.name}-oneof-{self.condition_branches_group._get_oneof_id()}'
+        )
+
+    def _validate_channels(
+        self,
+        channels: List[Union[PipelineParameterChannel,
+                             PipelineArtifactChannel]],
+    ):
+        self._validate_no_collected_channel(channels)
+        self._validate_no_oneof_channel(channels)
+        self._validate_no_mix_of_parameters_and_artifacts(channels)
+        self._validate_has_else_group(self.condition_branches_group)
+
+    def _validate_no_collected_channel(
+        self, channels: List[Union[PipelineParameterChannel,
+                                   PipelineArtifactChannel]]
+    ) -> None:
+        # avoid circular imports
+        from kfp.dsl import for_loop
+        if any(isinstance(channel, for_loop.Collected) for channel in channels):
+            raise ValueError(
+                f'dsl.{for_loop.Collected.__name__} cannot be used inside of dsl.{OneOf.__name__}.'
+            )
+
+    def _validate_no_oneof_channel(
+        self, channels: List[Union[PipelineParameterChannel,
+                                   PipelineArtifactChannel]]
+    ) -> None:
+        if any(isinstance(channel, OneOfMixin) for channel in channels):
+            raise ValueError(
+                f'dsl.{OneOf.__name__} cannot be used inside of another dsl.{OneOf.__name__}.'
+            )
+
+    def _validate_no_mix_of_parameters_and_artifacts(
+        self, channels: List[Union[PipelineParameterChannel,
+                                   PipelineArtifactChannel]]
+    ) -> None:
+        readable_name_map = {
+            PipelineParameterChannel: 'parameter',
+            PipelineArtifactChannel: 'artifact',
+            OneOfParameter: 'parameter',
+            OneOfArtifact: 'artifact',
+        }
+        # if channels[0] is any subclass of a PipelineParameterChannel
+        # check the rest of the channels against that parent type
+        # ensures check permits OneOfParameter and PipelineParameterChannel
+        # to be passed to OneOf together
+        if isinstance(channels[0], PipelineParameterChannel):
+            expected_type = PipelineParameterChannel
+        else:
+            expected_type = PipelineArtifactChannel
+
+        for i, channel in enumerate(channels[1:], start=1):
+            if not isinstance(channel, expected_type):
+                raise TypeError(
+                    f'Task outputs passed to dsl.{OneOf.__name__} must be the same type. Got two channels with different types: {readable_name_map[expected_type]} at index 0 and {readable_name_map[type(channel)]} at index {i}.'
+                )
+
+    def _validate_has_else_group(
+        self,
+        parent_group: 'tasks_group.ConditionBranches',
+    ) -> None:
+        # avoid circular imports
+        from kfp.dsl import tasks_group
+        if not isinstance(parent_group.groups[-1], tasks_group.Else):
```

Comment:
Yes, exactly.

For posterity, I'll leave a few notes to document my thinking:
- This is aligned with the current treatment of [`value_from_oneof`](https://github.com/kubeflow/pipelines/blob/1682ce8adeb2c55a155588eae7492b2f0a8b783a/api/v2alpha1/pipeline_spec.proto#L162-L165) field in `PipelineSpec` on Vertex Pipelines, which is a backend that supports such collection.
- **If we want different behavior...** We can, of course, prescribe new behavior for any backend and drop this validation in the KFP SDK. Since the current implementation is forward compatible with looser validation, my preference is to proceed with the strong validation and listen for user demands to loosen. This helps us get this feature in the hands of users faster with effectively no added cost for users or maintainers.
- **If we like this behavior, but want to permit more flexibility...** It's technically possible that a user might want to use only `dsl.If` and `dsl.Elif` and ensure the conditions are mutually exclusive themselves (no `dsl.Else`). Since this comes with the cost of shifting errors right (to runtime, instead of compilation time as in the current implementation) and because this seems like it wouldn't be the typical case, my preference is again to start conservatively with the current approach.